storagemarket/impl/providerstates/provider_states.go

```diff
@@ -220,32 +223,35 @@ func PublishDeal(ctx fsm.Context, environment ProviderDealEnvironment, deal stor
 
 // WaitForPublish waits for the publish message on chain and sends the deal id back to the client
 func WaitForPublish(ctx fsm.Context, environment ProviderDealEnvironment, deal storagemarket.MinerDeal) error {
-	return environment.Node().WaitForMessage(deal.PublishCid, storagemarket.ChainConfidence, func(code exitcode.ExitCode, retBytes []byte) error {
-		if code == exitcode.Ok {
-			var retval market.PublishStorageDealsReturn
-			err := retval.UnmarshalCBOR(bytes.NewReader(retBytes))
-			if err != nil {
-				return err
-			}
-
-			err = environment.SendSignedResponse(ctx.Context(), &network.Response{
-				State:          storagemarket.StorageDealProposalAccepted,
-				Proposal:       deal.ProposalCid,
-				PublishMessage: &deal.PublishCid,
-			})
-
-			if err != nil {
-				return ctx.Trigger(storagemarket.ProviderEventSendResponseFailed, err)
-			}
-
-			if err := environment.Disconnect(deal.ProposalCid); err != nil {
-				log.Warnf("closing client connection: %+v", err)
-			}
-
-			return ctx.Trigger(storagemarket.ProviderEventDealPublished, retval.IDs[0])
+	return environment.Node().WaitForMessage(ctx.Context(), deal.PublishCid, func(code exitcode.ExitCode, retBytes []byte, err error) error {
+		if err != nil {
+			return ctx.Trigger(storagemarket.ProviderEventDealPublishError, xerrors.Errorf("PublishStorageDeals errored: %w", err))
+		}
+		if code != exitcode.Ok {
+			return ctx.Trigger(storagemarket.ProviderEventDealPublishError, xerrors.Errorf("PublishStorageDeals exit code: %s", code.String()))
+		}
+		var retval market.PublishStorageDealsReturn
+		err = retval.UnmarshalCBOR(bytes.NewReader(retBytes))
+		if err != nil {
+			return ctx.Trigger(storagemarket.ProviderEventDealPublishError, xerrors.Errorf("PublishStorageDeals error unmarshalling result: %w", err))
+		}
+
+		err = environment.SendSignedResponse(ctx.Context(), &network.Response{
+			State:          storagemarket.StorageDealProposalAccepted,
+			Proposal:       deal.ProposalCid,
+			PublishMessage: &deal.PublishCid,
+		})
+
+		if err != nil {
+			return ctx.Trigger(storagemarket.ProviderEventSendResponseFailed, err)
 		}
 
-		return ctx.Trigger(storagemarket.ProviderEventDealPublishError, xerrors.Errorf("PublishStorageDeals exit code: %w", code))
+		if err := environment.Disconnect(deal.ProposalCid); err != nil {
```

Comment:
I'm curious about Disconnect. Why is Disconnect called only when is no error?  Also, there is a Disconnect call in SendSignedResponse, although in the tests it shows that it's harmless to do it twice.

Reply:
Disconnect gets called in the DealFailing handler, so no need to call it here. It probably should not be calling disconnect in SendSignedResponse and I will check that and add a ticket if needed.